### DISCUSSION TITLE
Prometheus custom step alignment feature

### DISCUSSION THREAD

**Discussion by fxmiii** [grafana/grafana#40208] [Data Sources]
**What would you like to be added**:  A feature that gives the ability to have a custom step alignment on prometheus panels. Specifically I want this feature for cases where I display prometheus counter changes over intervals of time. [A PR showing how it might work](https://github.com/grafana/grafana/pull/18148)

**Why is this needed**: Its needed for bar charts of prometheus counters with intervals that don't align nicely with midnight.

From the [PR](https://github.com/grafana/grafana/pull/18148):
> The classic case is where a facility runs work shifts (day shift and night shift) and the shift change time may not align with midnight. So for example:
> 
> Facility A operates with two shifts with shift change at 05:00 and 17:00. To display a nice bar chart of their production my query looks like "increase(ProductionCounts[12h])" with 'Min step'=12h and 'Step align'=5h
> 
> Facility B operates with three shifts with shift change at 7:00, 15:00, 23:00. A bar chart of their production my query looks like "increase(ProductionCounts[8h])" with 'Min step'=8h and 'Step align'=7h

**Comment by hugohaggmark** [comment]
Hi @fxmiii and thank you for creating this feature request. As Torkel mentioned we'll see the interest this gets.

**Comment by gabor** [comment]
there is the idea that if we can align the timerange, it should solve this for basically every datasource. we need to investigate:
- is this already possible with the timerange controls or not?
- will this solve the problem?

**Comment by ivanahuckova** [comment]
After discussion with @gabor we agreed that this is a feature request specifically for Prometheus data source and it doesn't make sense to solve it for other data sources. Basically issue is that when you run query such as `increase(go_gc_duration_seconds_count[24h])`, select `Last 7days` and update `step to 24h`. Each day starts with 00:00:00 time. On backend, we override time range and use aligned time range so the final results start at `00:00:00`. However, some users would prefer for it to start at other time - e.g.  `05:00:00` or `17:00:00` and this can be done only by new query input/param - custom step alignment.

I am moving this to github discussions.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
